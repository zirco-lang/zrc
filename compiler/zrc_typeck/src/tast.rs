//! Type-checked AST for the Zirco programming language
//!
//! This is similar to the [standard AST](zrc_parser::ast), but with type
//! information attached to it.
//!
//! Note that we only attach source locations to code that would exist after code generation,
//! meaning inlined types do not get a span.

pub mod expr;
pub mod stmt;
pub mod ty;
