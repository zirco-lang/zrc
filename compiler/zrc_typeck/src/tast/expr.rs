//! Expression representation for the Zirco [TAST](super)

use std::{fmt::Display, string::ToString};

pub use zrc_parser::{
    ast::expr::{Arithmetic, BinaryBitwise, Comparison, Equality, Logical},
    lexer::StringTok,
};

/// An [expression kind](TypedExprKind) with its yielded [result
/// type](super::ty::Type) attached to it.
#[derive(PartialEq, Debug, Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct TypedExpr<'input>(pub super::ty::Type<'input>, pub TypedExprKind<'input>);

/// The left hand side of an assignment.
#[derive(PartialEq, Debug, Clone)]
pub struct Place<'input>(pub super::ty::Type<'input>, pub PlaceKind<'input>);
/// The valid left-hand-side of a [`TypedExprKind::Assignment`].
///
/// Places may be:
/// - A variable or an property access of a place
/// - A dereference or index into any expression yielding a pointer
#[derive(PartialEq, Debug, Clone)]
pub enum PlaceKind<'input> {
    /// `*x`
    Deref(Box<TypedExpr<'input>>),
    /// `x`
    Variable(&'input str),
    /// `x[y]`
    Index(Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// `x.y`
    Dot(Box<Place<'input>>, &'input str),
}

/// The kind of a [`TypedExpr`]
#[derive(PartialEq, Debug, Clone)]
pub enum TypedExprKind<'input> {
    /// `a, b`
    Comma(Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),

    /// `a = b`
    ///
    /// `a += b` is desugared to `a = a + b` by the type checker.
    Assignment(Box<Place<'input>>, Box<TypedExpr<'input>>),

    /// Bitwise operations
    BinaryBitwise(
        BinaryBitwise,
        Box<TypedExpr<'input>>,
        Box<TypedExpr<'input>>,
    ),
    /// Logical operations
    Logical(Logical, Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// Equality checks
    Equality(Equality, Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// Comparisons
    Comparison(Comparison, Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// Arithmetic operations
    Arithmetic(Arithmetic, Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),

    /// `!x`
    UnaryNot(Box<TypedExpr<'input>>),
    /// `~x`
    UnaryBitwiseNot(Box<TypedExpr<'input>>),
    /// `-x`
    UnaryMinus(Box<TypedExpr<'input>>),
    /// `&x`
    UnaryAddressOf(Box<Place<'input>>),
    /// `*x`
    UnaryDereference(Box<TypedExpr<'input>>),

    /// `a[b]`
    Index(Box<TypedExpr<'input>>, Box<TypedExpr<'input>>),
    /// `a.b`
    Dot(Box<Place<'input>>, &'input str),
    /// `a(b, c, d, ...)`
    Call(Box<Place<'input>>, Vec<TypedExpr<'input>>),

    /// `a ? b : c`
    Ternary(
        Box<TypedExpr<'input>>,
        Box<TypedExpr<'input>>,
        Box<TypedExpr<'input>>,
    ),

    /// `x as T`
    Cast(Box<TypedExpr<'input>>, super::ty::Type<'input>),

    /// Any numeric literal.
    NumberLiteral(&'input str),
    /// Any string literal.bool
    StringLiteral(Vec<StringTok<'input>>),
    /// Any identifier.
    Identifier(&'input str),
    /// Any boolean literal.
    BooleanLiteral(bool),
}

impl<'input> Display for TypedExprKind<'input> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NumberLiteral(n) => write!(f, "{n}"),
            Self::StringLiteral(str) => write!(
                f,
                "\"{}\"",
                str.iter().map(ToString::to_string).collect::<String>()
            ),
            Self::Identifier(i) => write!(f, "{i}"),
            Self::BooleanLiteral(value) => write!(f, "{value}"),
            Self::Assignment(place, value) => write!(f, "{place} = {value}"),
            Self::Equality(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::Comparison(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::Arithmetic(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::BinaryBitwise(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::Logical(operator, lhs, rhs) => write!(f, "{lhs} {operator} {rhs}"),
            Self::Comma(lhs, rhs) => write!(f, "{lhs}, {rhs}"),
            Self::UnaryNot(expr) => write!(f, "!{expr}"),
            Self::UnaryBitwiseNot(expr) => write!(f, "~{expr}"),
            Self::UnaryMinus(expr) => write!(f, "-{expr}"),
            Self::UnaryAddressOf(expr) => write!(f, "&{expr}"),
            Self::UnaryDereference(expr) => write!(f, "*{expr}"),
            Self::Ternary(cond, if_true, if_false) => write!(f, "{cond} ? {if_true} : {if_false}"),
            Self::Index(ptr, idx) => write!(f, "{ptr}[{idx}]"),
            Self::Dot(expr, key) => write!(f, "{expr}.{key}"),
            Self::Cast(expr, ty) => write!(f, "{expr} as {ty}"),
            Self::Call(expr, args) => write!(
                f,
                "{expr}({})",
                args.iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
        }
    }
}
impl<'input> Display for PlaceKind<'input> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Deref(expr) => write!(f, "*{expr}"),
            Self::Variable(ident) => write!(f, "{ident}"),
            Self::Index(ptr, idx) => write!(f, "{ptr}[{idx}]"),
            Self::Dot(expr, key) => write!(f, "{expr}.{key}"),
        }
    }
}
impl<'input> Display for Place<'input> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "({}) as ({})", self.1, self.0)
    }
}

impl<'input> Display for TypedExpr<'input> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(({}) as ({}))", self.1, self.0)
    }
}
