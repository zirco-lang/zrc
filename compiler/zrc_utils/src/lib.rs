#![doc=include_str!("../README.md")]
#![allow(unknown_lints)] // in case you use non-nightly clippy
#![warn(
    clippy::cargo,
    clippy::nursery,
    clippy::pedantic,
    clippy::missing_docs_in_private_items,
    missing_docs,
    clippy::absolute_paths,
    clippy::as_conversions,
    clippy::dbg_macro,
    clippy::decimal_literal_representation,
    clippy::deref_by_slicing,
    clippy::disallowed_script_idents,
    clippy::else_if_without_else,
    clippy::empty_structs_with_brackets,
    clippy::format_push_string,
    clippy::if_then_some_else_none,
    clippy::let_underscore_must_use,
    clippy::min_ident_chars,
    clippy::mixed_read_write_in_expression,
    clippy::multiple_inherent_impl,
    clippy::multiple_unsafe_ops_per_block,
    clippy::non_ascii_literal,
    clippy::redundant_type_annotations,
    clippy::rest_pat_in_fully_bound_structs,
    clippy::same_name_method,
    clippy::semicolon_inside_block,
    clippy::unseparated_literal_suffix,
    clippy::string_to_string,
    clippy::todo,
    clippy::undocumented_unsafe_blocks,
    clippy::unimplemented,
    clippy::unneeded_field_pattern,
    clippy::wildcard_enum_match_arm,

    // These should be enabled in any non-user-facing code, like the parser, but not in the
    // frontend.
    clippy::print_stderr,
    clippy::print_stdout
)]
#![allow(clippy::multiple_crate_versions, clippy::cargo_common_metadata)]

pub mod span;
