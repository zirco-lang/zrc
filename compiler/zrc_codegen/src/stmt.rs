//! Code generation for statements

use inkwell::{
    basic_block::BasicBlock,
    debug_info::{AsDIScope, DILexicalBlock},
};
use zrc_typeck::tast::{
    expr::{Place, PlaceKind, TypedExpr, TypedExprKind},
    stmt::{LetDeclaration, TypedStmt, TypedStmtKind},
    ty::Type,
};
use zrc_utils::span::{Span, Spannable, Spanned};

use crate::{
    ctx::{BlockCtx, FunctionCtx},
    expr::cg_expr,
    ty::llvm_basic_type,
    BasicBlockAnd, CgScope,
};

/// Consists of the [`BasicBlock`]s to `br` to when encountering certain
/// instructions. It is passed to [`cg_block`] to allow it to properly handle
/// break and continue.
#[derive(PartialEq, Eq, Debug, Clone)]
#[allow(clippy::redundant_pub_crate)]
pub(crate) struct LoopBreakaway<'ctx> {
    /// Points to the exit basic block.
    on_break: BasicBlock<'ctx>,
    /// For `for` loops, points to the latch. For `while` loops, points to the
    /// header.
    on_continue: BasicBlock<'ctx>,
}

/// Generates the `alloca`tion, `store` instruction, and adds a new identifier
/// to the [`CgScope`].
///
/// # Panics
/// Panics if an internal code generation error is encountered.
#[allow(clippy::too_many_arguments)]
fn cg_let_declaration<'ctx, 'input, 'a>(
    cg: FunctionCtx<'ctx, 'a>,
    mut bb: BasicBlock<'ctx>,
    scope: &'a mut CgScope<'input, 'ctx>,
    dbg_scope: DILexicalBlock<'ctx>,
    declarations: Vec<Spanned<LetDeclaration<'input>>>,
) -> BasicBlock<'ctx> {
    for spanned_let_declaration in declarations {
        let span = spanned_let_declaration.span();
        let let_declaration = spanned_let_declaration.into_value();

        let stmt_line_col = cg.line_lookup.lookup_from_index(span.start());
        let debug_location = cg.dbg_builder.create_debug_location(
            cg.ctx,
            stmt_line_col.line,
            stmt_line_col.col,
            dbg_scope.as_debug_info_scope(),
            None,
        );
        cg.builder.set_current_debug_location(debug_location);

        // we create our own builder here because we need to insert the alloca
        // at the beginning of the entry block, and that is easier than trying to
        // somehow save our position.

        let entry_block_builder = cg.ctx.create_builder();
        let first_bb = cg
            .fn_value
            .get_first_basic_block()
            .expect("function should have at least one basic block");

        #[allow(clippy::option_if_let_else)]
        match first_bb.get_first_instruction() {
            Some(first_instruction) => {
                entry_block_builder.position_before(&first_instruction);
            }
            None => {
                entry_block_builder.position_at_end(first_bb);
            }
        }

        let (ty, dbg_ty) = llvm_basic_type(&cg, &let_declaration.ty);

        let ptr = entry_block_builder
            .build_alloca(ty, &format!("let_{}", let_declaration.name))
            .expect("alloca should generate successfully");

        scope.insert(let_declaration.name.value(), ptr);

        let decl = cg.dbg_builder.create_auto_variable(
            dbg_scope.as_debug_info_scope(),
            let_declaration.name.value(),
            cg.compilation_unit.get_file(),
            cg.line_lookup.lookup_from_index(span.start()).line,
            dbg_ty,
            true,
            0,
            0,
        );

        cg.dbg_builder
            .insert_declare_at_end(ptr, Some(decl), None, debug_location, first_bb);

        if let Some(value) = let_declaration.value {
            let expr_cg = BlockCtx::new(cg, scope, dbg_scope);

            bb = cg_expr(
                expr_cg,
                bb,
                TypedExpr {
                    inferred_type: let_declaration.ty.clone(),
                    kind: value.kind.span().containing(TypedExprKind::Assignment(
                        Box::new(Place {
                            inferred_type: let_declaration.ty,
                            kind: PlaceKind::Variable(let_declaration.name.value())
                                .in_span(let_declaration.name.span()),
                        }),
                        Box::new(value),
                    )),
                },
            )
            .bb;
        }
    }

    bb
}

/// Process a vector of [`TypedStmt`]s (a block) and handle each statement.
///
/// # Panics
/// Panics if an internal code generation error is encountered.
#[allow(
    clippy::too_many_arguments,
    clippy::too_many_lines,
    clippy::needless_pass_by_value,
    clippy::redundant_pub_crate
)]
pub(crate) fn cg_block<'ctx, 'input, 'a>(
    cg: FunctionCtx<'ctx, 'a>,
    bb: BasicBlock<'ctx>,
    parent_scope: &'a CgScope<'input, 'ctx>,
    parent_lexical_block: DILexicalBlock<'ctx>,
    block: Spanned<Vec<TypedStmt<'input>>>,
    breakaway: &Option<LoopBreakaway<'ctx>>,
) -> Option<BasicBlock<'ctx>> {
    let mut scope = parent_scope.clone();
    let block_span = block.span();
    let block_line_col = cg.line_lookup.lookup_from_index(block_span.start());
    let lexical_block = cg.dbg_builder.create_lexical_block(
        parent_lexical_block.as_debug_info_scope(),
        cg.compilation_unit.get_file(),
        block_line_col.line,
        block_line_col.col,
    );

    block
        .into_value()
        .into_iter()
        .try_fold(bb, |bb, stmt| -> Option<BasicBlock> {
            let stmt_span = stmt.0.span();
            let stmt_line_col = cg.line_lookup.lookup_from_index(stmt_span.start());
            let debug_location = cg.dbg_builder.create_debug_location(
                cg.ctx,
                stmt_line_col.line,
                stmt_line_col.col,
                lexical_block.as_debug_info_scope(),
                None,
            );
            cg.builder.set_current_debug_location(debug_location);

            match stmt.0.into_value() {
                TypedStmtKind::ExprStmt(expr) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    Some(cg_expr(expr_cg, bb, expr).bb)
                }

                TypedStmtKind::IfStmt(cond, then, then_else) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    let then_else = then_else.unwrap_or_else(|| {
                        vec![].in_span(Span::from_positions(then.end(), then.end()))
                    });

                    let then_end = then.end();
                    let then_else_end = then_else.end();

                    let cond = cg_expr(expr_cg, bb, cond).into_value();

                    let then_bb = cg.ctx.append_basic_block(cg.fn_value, "then");
                    let then_else_bb = cg.ctx.append_basic_block(cg.fn_value, "then_else");

                    cg.builder
                        .build_conditional_branch(cond.into_int_value(), then_bb, then_else_bb)
                        .expect("conditional branch should generate successfully");

                    cg.builder.position_at_end(then_bb);
                    let maybe_then_bb =
                        cg_block(cg, then_bb, &scope, lexical_block, then, breakaway);

                    cg.builder.position_at_end(then_else_bb);
                    let maybe_then_else_bb = cg_block(
                        cg,
                        then_else_bb,
                        &scope,
                        lexical_block,
                        then_else,
                        breakaway,
                    );

                    match (maybe_then_bb, maybe_then_else_bb) {
                        (None, None) => None,
                        (Some(single_bb), None) | (None, Some(single_bb)) => {
                            let end = cg.ctx.append_basic_block(cg.fn_value, "end");

                            let then_end_line_col = cg.line_lookup.lookup_from_index(then_end);
                            let terminating_debug_location = cg.dbg_builder.create_debug_location(
                                cg.ctx,
                                then_end_line_col.line,
                                then_end_line_col.col,
                                lexical_block.as_debug_info_scope(),
                                None,
                            );

                            cg.builder
                                .set_current_debug_location(terminating_debug_location);

                            cg.builder.position_at_end(single_bb);
                            cg.builder
                                .build_unconditional_branch(end)
                                .expect("branch should generate successfully");

                            cg.builder.position_at_end(end);
                            Some(end)
                        }
                        (Some(then_bb), Some(then_else_bb)) => {
                            let end = cg.ctx.append_basic_block(cg.fn_value, "end");

                            let then_end_line_col = cg.line_lookup.lookup_from_index(then_end);
                            let then_terminating_debug_location =
                                cg.dbg_builder.create_debug_location(
                                    cg.ctx,
                                    then_end_line_col.line,
                                    then_end_line_col.col,
                                    lexical_block.as_debug_info_scope(),
                                    None,
                                );
                            cg.builder
                                .set_current_debug_location(then_terminating_debug_location);
                            cg.builder.position_at_end(then_bb);
                            cg.builder
                                .build_unconditional_branch(end)
                                .expect("branch should generate successfully");

                            let then_else_end_line_col =
                                cg.line_lookup.lookup_from_index(then_else_end);
                            let then_else_terminating_debug_location =
                                cg.dbg_builder.create_debug_location(
                                    cg.ctx,
                                    then_else_end_line_col.line,
                                    then_else_end_line_col.col,
                                    lexical_block.as_debug_info_scope(),
                                    None,
                                );
                            cg.builder
                                .set_current_debug_location(then_else_terminating_debug_location);

                            cg.builder.position_at_end(then_else_bb);
                            cg.builder
                                .build_unconditional_branch(end)
                                .expect("branch should generate successfully");

                            cg.builder.position_at_end(end);
                            Some(end)
                        }
                    }
                }

                TypedStmtKind::BlockStmt(block) => cg_block(
                    cg,
                    bb,
                    &scope,
                    lexical_block,
                    block.in_span(stmt_span),
                    breakaway,
                ),

                TypedStmtKind::ReturnStmt(Some(expr)) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    let expr = cg_expr(expr_cg, bb, expr).into_value();

                    cg.builder
                        .build_return(Some(&expr))
                        .expect("return should generate successfully");

                    None
                }

                TypedStmtKind::ReturnStmt(None) => {
                    let unit_type = llvm_basic_type(&cg, &Type::unit());

                    cg.builder
                        .build_return(Some(&unit_type.0.const_zero()))
                        .expect("return should generate successfully");

                    None
                }

                TypedStmtKind::ContinueStmt => {
                    cg.builder
                        .build_unconditional_branch(
                            breakaway
                                .as_ref()
                                .expect("`breakaway` should exist all places `continue` is valid")
                                .on_continue,
                        )
                        .expect("branch should generate successfully");

                    None
                }

                TypedStmtKind::BreakStmt => {
                    cg.builder
                        .build_unconditional_branch(
                            breakaway
                                .as_ref()
                                .expect("`breakaway` should exist all places `break` is valid")
                                .on_break,
                        )
                        .expect("branch should generate successfully");

                    None
                }

                TypedStmtKind::DeclarationList(declarations) => Some(cg_let_declaration(
                    cg,
                    bb,
                    &mut scope,
                    lexical_block,
                    declarations,
                )),

                TypedStmtKind::ForStmt {
                    init,
                    cond,
                    post,
                    body,
                } => {
                    // For loops generate a somewhat more complicated CFG, with a few parts.
                    // The preheader, where `init` runs. Breaks to the header.
                    // The header, where `cond` is checked and breaks to either the exit or the
                    // body. The body, where most of the body runs. Breaks to
                    // the latch. `break` transfers to the exit by force and `continue` transfers to
                    // the latch by force. The latch, where `post` runs and
                    // breaks back to the header The exit, which is the basic
                    // block we return.

                    // loops lie in an implicit subscope
                    let mut scope = scope.clone();

                    // The block we are currently in will become the preheader. Generate the `init`
                    // code if there is any.
                    if let Some(init) = init {
                        cg_let_declaration(cg, bb, &mut scope, lexical_block, *init);
                    }

                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    let header = cg.ctx.append_basic_block(cg.fn_value, "header");
                    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");
                    let latch = cg.ctx.append_basic_block(cg.fn_value, "latch");
                    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

                    // Branch to the header from the preheader.
                    cg.builder
                        .build_unconditional_branch(header)
                        .expect("branch should generate successfully");

                    // Generate the header.
                    cg.builder.position_at_end(header);
                    let header = cond.map_or_else(
                        || {
                            // If there is no condition, we always branch to the body.
                            cg.builder
                                .build_unconditional_branch(body_bb)
                                .expect("branch should generate successfully");

                            header
                        },
                        |cond| {
                            let mut header = header;

                            let cond = unpack!(header = cg_expr(expr_cg, header, cond));

                            cg.builder
                                .build_conditional_branch(cond.into_int_value(), body_bb, exit)
                                .expect("branch should generate successfully");

                            header
                        },
                    );

                    // Generate the body.
                    cg.builder.position_at_end(body_bb);
                    let body_bb = cg_block(
                        cg,
                        body_bb,
                        &scope,
                        lexical_block,
                        body,
                        &Some(LoopBreakaway {
                            on_break: exit,
                            on_continue: latch,
                        }),
                    );

                    // The body breaks to latch
                    if body_bb.is_some() {
                        cg.builder
                            .build_unconditional_branch(latch)
                            .expect("branch should generate successfully");
                    }

                    // Latch runs post and then breaks right back to the header.
                    cg.builder.position_at_end(latch);
                    if let Some(post) = post {
                        cg_expr(expr_cg, latch, post);
                    }

                    cg.builder
                        .build_unconditional_branch(header)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(exit);

                    Some(exit)
                }

                TypedStmtKind::WhileStmt(cond, body) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    // While loops are similar to for loops but much simpler.
                    // The preheader simply just breaks to the header.
                    // The header checks the condition and breaks to the exit or the body.
                    // The body simply breaks to the header.
                    // The exit is the continued code

                    // `break` => exit
                    // `continue` => header

                    let mut header = cg.ctx.append_basic_block(cg.fn_value, "header");

                    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");

                    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

                    cg.builder
                        .build_unconditional_branch(header)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(header);

                    let cond = unpack!(header = cg_expr(expr_cg, header, cond));

                    cg.builder
                        .build_conditional_branch(cond.into_int_value(), body_bb, exit)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(body_bb);

                    let body_bb = cg_block(
                        cg,
                        body_bb,
                        &scope,
                        lexical_block,
                        body,
                        &Some(LoopBreakaway {
                            on_break: exit,
                            on_continue: header,
                        }),
                    );

                    if body_bb.is_some() {
                        cg.builder
                            .build_unconditional_branch(header)
                            .expect("branch should generate successfully");
                    }

                    cg.builder.position_at_end(exit);

                    Some(exit)
                }

                TypedStmtKind::DoWhileStmt(body, cond) => {
                    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

                    // `do..while` loops are slightly different from `while` loops.
                    // the preheader breaks directly to the *body* and forces it to run at
                    // least once. the body can then later break to the header which checks the
                    // condition and will loop or exit.

                    // `break` => exit
                    // `continue` => header

                    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");
                    let body_start = body_bb;

                    let header = cg.ctx.append_basic_block(cg.fn_value, "header");

                    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

                    cg.builder
                        .build_unconditional_branch(body_bb)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(body_bb);

                    let body_bb = cg_block(
                        cg,
                        body_bb,
                        &scope,
                        lexical_block,
                        body,
                        &Some(LoopBreakaway {
                            on_break: exit,
                            on_continue: header,
                        }),
                    );

                    if body_bb.is_some() {
                        cg.builder
                            .build_unconditional_branch(header)
                            .expect("branch should generate successfully");
                    }

                    cg.builder.position_at_end(header);

                    let BasicBlockAnd { value: cond, .. } = cg_expr(expr_cg, header, cond);

                    cg.builder
                        .build_conditional_branch(cond.into_int_value(), body_start, exit)
                        .expect("branch should generate successfully");

                    cg.builder.position_at_end(exit);

                    Some(exit)
                }
            }
        })
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn let_declarations_are_properly_generated() {
        cg_snapshot_test!(indoc! {"
            fn test() {
                // TEST: should allocate twice and assign to one.
                let a: i32;
                let b: i32 = 7;
            }
        "});
    }

    mod cg_block {
        use super::*;

        #[test]
        fn function_parameters_are_properly_generated() {
            cg_snapshot_test!(indoc! {"
                fn id(x: i32) -> i32 {
                    return x;
                }
            "});
        }

        mod conditionals {
            use super::*;

            #[test]
            fn if_statements_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;
                    fn nop();

                    fn test() {
                        // TEST: properly produces a conditional break over the call result and
                        // both code paths join at the end
                        if (get_bool()) nop();

                        // TEST: code generation properly continues in the last block
                        nop();
                        return;
                    }
                "});
            }

            #[test]
            fn if_else_statements_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;
                    fn nop();

                    fn test() {
                        // TEST: properly produces a conditional break over the call result and
                        // both code baths call nop().
                        if (get_bool()) nop();
                        else {
                            nop();
                            // TEST: this path diverges
                            return;
                        }

                        // TEST: code generation properly continues from the if_true block
                        nop();
                    }
                "});
            }

            #[test]
            fn if_else_statements_where_both_blocks_terminate_do_not_continue_generating() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: properly produces a conditional break over the call result
                        // and both code paths return (diverge). there should be no %end bb.
                        if (get_bool()) return;
                        else return;
                    }
                "});
            }
        }

        mod loops {
            use super::*;

            #[test]
            fn while_loops_along_with_break_and_continue_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: the proper while loop structure is created
                        while (get_bool()) {
                            // TEST: break jumps to the `end` block
                            if (get_bool()) break;
                            else {
                                // TEST: continue jumps to the header block
                                if (get_bool()) continue;
                                // TEST: otherwise, we proceed
                                else {}
                            }

                            // TEST: the loop jumps back to the header block
                        }

                        // TEST: ...and code generation properly continues.
                        return;
                    }
                "});
            }

            #[test]
            fn for_loops_along_with_break_and_continue_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_int() -> i32;

                    fn test() {
                        // TEST: the proper while loop structure is created
                        for (let i = 0; i < get_int(); i += 1) {
                            // TEST: break jumps to the `end` block
                            if (i > get_int()) break;
                            else {
                                // TEST: continue jumps to the latch block
                                if (i < get_int()) continue;
                                else {}
                            }

                            // TEST: the loop jumps to the latch block which jumps back to the
                            // header
                        }
                    }
                "});
            }

            #[test]
            fn do_while_loops_generate_as_expected() {
                cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: the proper `do..while` loop structure is created
                        do {
                            get_bool(); // for fake side effects
                        } while (get_bool());
                    }
                "});
            }
        }
    }
}
