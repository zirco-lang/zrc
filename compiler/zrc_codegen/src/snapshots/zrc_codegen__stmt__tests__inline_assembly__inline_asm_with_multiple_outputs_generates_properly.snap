---
source: compiler/zrc_codegen/src/stmt.rs
description: "fn test() {\n    let a: i32 = 10;\n    let b: i32 = 20;\n    let sum: i32 = 0;\n    let diff: i32 = 0;\n    // TEST: inline assembly with multiple output operands\n    asm(\"movl $2, $0; subl $3, $1\"\n        : \"=r\"(sum), \"=r\"(diff)\n        : \"r\"(a), \"r\"(b));\n}\n"
expression: resulting_ir
snapshot_kind: text
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

define {} @test() !dbg !3 {
entry:
  %let_diff = alloca i32, align 4
  %let_sum = alloca i32, align 4
  %let_b = alloca i32, align 4
  %let_a = alloca i32, align 4
  call void @llvm.dbg.declare(metadata ptr %let_a, metadata !9, metadata !DIExpression()), !dbg !16
  store i32 10, ptr %let_a, align 4, !dbg !16
  call void @llvm.dbg.declare(metadata ptr %let_b, metadata !13, metadata !DIExpression()), !dbg !17
  store i32 20, ptr %let_b, align 4, !dbg !17
  call void @llvm.dbg.declare(metadata ptr %let_sum, metadata !14, metadata !DIExpression()), !dbg !18
  store i32 0, ptr %let_sum, align 4, !dbg !18
  call void @llvm.dbg.declare(metadata ptr %let_diff, metadata !15, metadata !DIExpression()), !dbg !19
  store i32 0, ptr %let_diff, align 4, !dbg !19
  %load = load i32, ptr %let_a, align 4, !dbg !20
  %load1 = load i32, ptr %let_b, align 4, !dbg !21
  %inline_asm = call { i32, i32 } asm sideeffect "movl $2, $0; subl $3, $1", "=r,=r,r,r"(i32 %load, i32 %load1), !dbg !21
  %out0 = extractvalue { i32, i32 } %inline_asm, 0, !dbg !21
  store i32 %out0, ptr %let_sum, align 4, !dbg !21
  %out1 = extractvalue { i32, i32 } %inline_asm, 1, !dbg !21
  store i32 %out1, ptr %let_diff, align 4, !dbg !21
  ret {} zeroinitializer, !dbg !22
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare void @llvm.dbg.declare(metadata, metadata, metadata) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 1, type: !4, scopeLine: 1, spFlags: DISPFlagDefinition, unit: !1, retainedNodes: !8)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DICompositeType(tag: DW_TAG_structure_type, name: "struct {}", scope: !2, file: !2, elements: !7)
!7 = !{}
!8 = !{!9, !13, !14, !15}
!9 = !DILocalVariable(name: "a", scope: !10, file: !2, line: 2, type: !12)
!10 = distinct !DILexicalBlock(scope: !11, file: !2, line: 1, column: 11)
!11 = distinct !DILexicalBlock(scope: !3, file: !2, line: 1, column: 11)
!12 = !DIBasicType(name: "i32")
!13 = !DILocalVariable(name: "b", scope: !10, file: !2, line: 3, type: !12)
!14 = !DILocalVariable(name: "sum", scope: !10, file: !2, line: 4, type: !12)
!15 = !DILocalVariable(name: "diff", scope: !10, file: !2, line: 5, type: !12)
!16 = !DILocation(line: 2, column: 9, scope: !10)
!17 = !DILocation(line: 3, column: 9, scope: !10)
!18 = !DILocation(line: 4, column: 9, scope: !10)
!19 = !DILocation(line: 5, column: 9, scope: !10)
!20 = !DILocation(line: 9, column: 15, scope: !10)
!21 = !DILocation(line: 9, column: 23, scope: !10)
!22 = !DILocation(line: 10, column: 1, scope: !10)
